### PR TITLE
Implementing Golod complexes

### PR DESCRIPTION
**Pull request by OP5642**
### Implementing Golod (simplicial) complexes 

This is a part of #35640 (GSoC 20223), relies on #35430 (bigraded Betti numbers).

Added two methods: `is_golod()` and `is_minimally_non_golod()`, defined for simplicial complexes. These are significant invariants used in toric topology.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by OP5642** [comment]
@tscrim

File: src/sage/topology/simplicial_complex.py [hunk 1]
```diff
@@ -4950,6 +4950,65 @@ def bigraded_betti_number(self, a, b, base_ring=ZZ):
...
+            sage: Y.is_golod()
+            True
+        """
+        H = set(a+b for (a, b) in self.bigraded_betti_numbers())
+
+        for i in H:
+            for j in H:
+                if i+j in H and i != 0 and j != 0:
+                    return False
+        return True
```
Comment (tscrim) [inline src/sage/topology/simplicial_complex.py]:
```suggestion
        H = set(a+b for (a, b) in self.bigraded_betti_numbers())
        if 0 in H:
            H.remove(0)

        return all(i + j in H for i in H for j in H)
```

File: src/sage/topology/simplicial_complex.py [hunk 2]
```diff
@@ -4950,6 +4950,65 @@ def bigraded_betti_number(self, a, b, base_ring=ZZ):
...
+            False
+        """
+        L = self.vertices()
+        n = len(L)
+
+        for x in combinations(L, n-1):
+            if not self.generated_subcomplex(x).is_golod():
+                return False
+
+        return True
```
Comment (tscrim) [inline src/sage/topology/simplicial_complex.py]:
```suggestion
        def test(v):
            X = copy(self)
            X.delete_vertex(v)
            return X.is_golod()

        return (not self.is_golod()) and all(test(v) for v in self.vertices())
```

**Review by tscrim — CHANGES_REQUESTED** [review]
Here's some changes that I think will make the code faster.

File: src/sage/topology/simplicial_complex.py [hunk 3]
```diff
@@ -4950,6 +4950,61 @@ def bigraded_betti_number(self, a, b, base_ring=ZZ):
...
+            sage: X.is_golod()
+            False
+            sage: Y.is_golod()
+            True
+        """
+        H = set(a+b for (a, b) in self.bigraded_betti_numbers())
+        if 0 in H:
+            H.remove(0)
+
+        return not any(i + j in H for i in H for j in H)
```
Comment (tscrim) [inline src/sage/topology/simplicial_complex.py]: I think I changed the logic slightly for `is_golod`. Before, it was possible that `i + j == 0`, which would have returned `False`. However, the current code doesn't catch this. Also, with the commutativity, we can simplify it.
```suggestion
        H = set(a+b for (a, b) in self.bigraded_betti_numbers())
        if 0 in H:
            HL = list(H.difference([0]))
        else:
            HL = list(H)

        return not any(i + j in H for ii, i in enumerate(HL) for j in HL[ii:])
```

**Review by tscrim — CHANGES_REQUESTED** [review]

**Review by tscrim — APPROVED** [review]
LGTM.

**Pull request merged**